Delete submissions and replies when a source is deleted

redshiftzero opened the pull request:
This PR is to merge in the work of @pwplus from #1190 which deletes submissions and replies when the source is deleted. I just added a unit test. Closes #1188.

conorsch commented on securedrop/tests/test_unit_journalist.py:
```diff
@@ -4,7 +4,7 @@
 import unittest
 import zipfile
-
+import pdb
```
Remove `pdb` import left over from debugging.

garrettr commented on securedrop/db.py:
```diff
@@ -132,7 +132,8 @@ class Submission(Base):
             'submissions',
-            order_by=id))
+            order_by=id,
+        cascade="delete"))
```
Nit: formatting. Either put all of the parameter to `backref` on one line, or left-align them. If you can fit them all of one line, that is preferable.

redshiftzero replied:
Fixed!

garrettr commented on securedrop/tests/test_unit_journalist.py:
```diff
@@ -208,18 +217,29 @@ def test_edit_hotp(self):
+        # Submissions should be gone
+        results = db_session.query(Submission.source_id == source.id).all()
+        self.assertEqual(results, [])
+
```
I think you should also create a test reply as well, and verify that it is deleted as well (in order to test the change to `class Reply`).

redshiftzero replied:
Ah, good point - I have now written this in a separate unit test (and added a helper function to setup test replies in the database).

redshiftzero commented:
Thanks for your review @garrettr! Let me know if these changes don't address your concerns.

redshiftzero commented on securedrop/db.py:
```diff
@@ -158,7 +158,10 @@ class Reply(Base):
 
     source_id = Column(Integer, ForeignKey('sources.id'))
-    source = relationship("Source", backref=backref('replies', order_by=id))
+    source = relationship(
```
I agree with your points here, and I definitely agree with you that the first is more readable than the second.  I’ll note that: 

```
    source = relationship("Source",
                          backref=backref("replies", order_by=id, cascade="delete"))
```

is not PEP8 compliant due to limits on characters per line. Instead I decided to opt for the ‘hanging intent’ style used in many of the tests in #1417 for readability. Happy to discuss further. 

Also: in order to address these kinds of style issues in the codebase, I suggest we tackle #47 and #886 soon and define and clean up the style in a few style-only PRs (I’m not going to change the lines you mentioned above in this PR since they’re not in the diff and I don’t want to confuse the history here, but the lines you point out are definitely something for a style PR once we’ve defined what constitutes good style for this project.)

redshiftzero commented:
So it looks like the PGP key is indeed being deleted when the source is deleted (done in `journalist.delete_collection`) but the deletion of the source documents can take a second due to the Redis worker. However I’ve added additional unit tests to explicitly check this behavior. 

Check it out and let me know if this doesn’t address your concerns @fowlslegs! Also rebased on current develop.

fowlslegs commented:
More good stuff @pwplus @redshiftzero!! Merging!
